Feature/str pattern completion

Ashwolaa opened the pull request:
Pattern completer subpart of PR#105:

This PR contains:
- the **PatternCompleter** class which can be used with QTextEdit/QLineEdit/QPlainTextEdit, resulting in PatternTextEdit/PatternLineEdit/PatternPlainTextEdit
See _pattern_completer_demo.py_ in examples for some demonstration.
- the **PatternParameter** which makes use of the PatternCompleter class. It provides a QPlainTextEdit making it similar to the usual parameter item but with the PatternCompleter methods.
See _parameter_ex.py_ in examples for an application

Ashwolaa committed:
pattern_completer_widget

Ashwolaa committed:
adding demo

Ashwolaa committed:
using qtpy

Ashwolaa committed:
adding text_pattern

Ashwolaa committed:
adding item to init

Ashwolaa committed:
updating parameter_ex

Ashwolaa committed:
Merge remote-tracking branch 'origin/bugfix/parameter_manager_losing_name_from_collapse' into feature/str_pattern_completion

Ashwolaa committed:
wrong title/name in example

Ashwolaa committed:
adding pattern and completer config to ioxml

Ashwolaa committed:
adding the elt to dict for pattern

Ashwolaa committed:
adding some tests

Ashwolaa committed:
renaming and moving test files

Ashwolaa committed:
updating title in example to be more self explanatory

Ashwolaa committed:
using explicit name in parameter_ex

Ashwolaa committed:
Merge branch 'PyMoDAQ:5.0.x_dev' into feature/str_pattern_completion

Ashwolaa committed:
fixing import in demo

Ashwolaa committed:
Merge branch 'feature/str_pattern_completion' of https://github.com/Ashwolaa/pymodaq_gui into feature/str_pattern_completion

seb5g commented:
A few comments to start on this huge PR:

* when there is only one elt in the proposed choices (from the pattern filtering) it would be nice to just press enter to use it. As it is, one has to use the down arrow key or the mouse click
* for emoji, you should include a space just after the emoji otherwise it breaks the emoji representation
* could not see any difference in your wordwrap example
*

Ashwolaa committed:
Adding first selection as default + highlighting it

Ashwolaa committed:
Adding more complex scheme (: vs ::)

Ashwolaa committed:
updating the demo

Ashwolaa committed:
updating label

Ashwolaa commented:
> A few comments to start on this huge PR:
> 
> * when there is only one elt in the proposed choices (from the pattern filtering) it would be nice to just press enter to use it. As it is, one has to use the down arrow key or the mouse click

I agree, I changed it to have a behavior close to an IDE. The first item proposed is selected by default. Pressing Enter/Tab will autocomplete it (Tab only works well with QPlainText though)

> * for emoji, you should include a space just after the emoji otherwise it breaks the emoji representation

I removed them and I changed the example with more standard Unicodes. The issue with the emojis is that they consist of multiple characters which make the whole identification struggle.

> * could not see any difference in your wordwrap example

I updated the example, it should be more explanatory now.

seb5g commented:
@malik-irain could you have a go to check this as well and merge it if it is fine. We have to move on to the monorepo ...

malik-irain commented on src/pymodaq_gui/parameter/pymodaq_ptypes/text_pattern.py:
```diff
@@ -0,0 +1,484 @@
+            pattern_config = self.param.opts.get(f"pattern_config_{pattern}", {})
+            widget.add_completer(pattern, completions, **pattern_config)
+
+        widget.value = widget.toPlainText
+        widget.setValue = widget.setPlainText
+        widget.sigChanged = widget.textChanged
```
I know that's what's the doc says about subclassing WidgetParameterItem but I would be careful about monkeypatching qt methods as it already made tests fails without any apparent reason.  I would have implemented a value attribute and a `setValue` method in `PatternPlainTextEdit` widget that call `toPlainText` and `setPlainText`. For the signal, it might be tricky.

Anyways for now, it works so we can leave it like that

Ashwolaa replied:
I can add this if you think it is better. There is no impact on the tests.

Not sure, how to deal with the signal though.

Ashwolaa replied:
Claude calls it "Duck typing adapter"

malik-irain replied:
Yeah it works very well in python, but when using objects from C++ it can lead to segmentation faults (we had the issue recently), but as it doesn't seem to make tests fail, it's ok with me!

malik-irain commented:
@Ashwolaa Can you rebase or merge the last modification of pymodaq_gui in you PR so that I can see if tests are passing?

Ashwolaa committed:
Merge branch 'PyMoDAQ:5.0.x_dev' into feature/str_pattern_completion
